server/src-lib/Hasura/RQL/Types/ScheduledTrigger.hs

```diff
@@ -166,3 +187,173 @@ instance FromJSON CreateScheduledEvent where
+  { _sepLimit  :: !(Maybe Int)
+  , _sepOffset :: !(Maybe Int)
+  } deriving (Show, Eq)
+
+parseScheduledEventPagination :: Object -> Parser ScheduledEventPagination
+parseScheduledEventPagination o =
+  ScheduledEventPagination
+    <$> o .:? "limit"
+    <*> o .:? "offset"
+
+scheduledEventPaginationToPairs :: ScheduledEventPagination -> [Pair]
+scheduledEventPaginationToPairs ScheduledEventPagination{..} =
+  ["limit" .= _sepLimit, "offset" .= _sepOffset]
+
+data GetInvocationsBy
+  = GIBEventId !EventId !ScheduledEventType
+  | GIBEvent !ScheduledEvent
+  deriving (Show, Eq)
+
+data GetEventInvocations
+  = GetEventInvocations
+  { _geiInvocationsBy :: !GetInvocationsBy
+  , _geiPagination    :: !ScheduledEventPagination
+  } deriving (Eq, Show)
+
+instance FromJSON GetEventInvocations where
+  parseJSON = withObject "Object" $ \o ->
+    GetEventInvocations
+      <$> (parseEventId o <|> (GIBEvent <$> parseScheduledEvent o))
+      <*> parseScheduledEventPagination o
+    where
+      parseEventId o =
+        GIBEventId <$> o .: "event_id" <*> o .: "type"
+
+instance ToJSON GetEventInvocations where
+  toJSON GetEventInvocations{..} =
+    object $ case _geiInvocationsBy of
+             GIBEventId eventId eventType -> ["event_id" .= eventId, "type" .= eventType]
+             GIBEvent event               -> scheduledEventToPairs event
+          <> scheduledEventPaginationToPairs _geiPagination
+
+data CronEventSeed
+  = CronEventSeed
+  { cesName          :: !TriggerName
+  , cesScheduledTime :: !UTCTime
+  } deriving (Show, Eq)
+
+data ScheduledEventSeed
+  = SESCron ![CronEventSeed]
+  | SESOneOff !CreateScheduledEvent
+  deriving (Show, Eq)
+
+data ScheduledEventStatus
+  = SESScheduled
+  | SESLocked
+  | SESDelivered
+  | SESError
+  | SESDead
+  deriving (Show, Eq)
+
+scheduledEventStatusToText :: ScheduledEventStatus -> Text
+scheduledEventStatusToText SESScheduled = "scheduled"
+scheduledEventStatusToText SESLocked    = "locked"
+scheduledEventStatusToText SESDelivered = "delivered"
+scheduledEventStatusToText SESError     = "error"
+scheduledEventStatusToText SESDead      = "dead"
+
+textToScheduledEventStatus :: Text -> Maybe ScheduledEventStatus
+textToScheduledEventStatus = \case
+  "scheduled" -> Just SESScheduled
+  "locked"    -> Just SESLocked
+  "delivered" -> Just SESDelivered
+  "error"     -> Just SESError
+  "dead"      -> Just SESDead
+  _           -> Nothing
+
+instance Q.ToPrepArg ScheduledEventStatus where
+  toPrepVal = Q.toPrepVal . scheduledEventStatusToText
+
+instance Q.FromCol ScheduledEventStatus where
+  fromCol bs =
+    flip Q.fromColHelper bs $ PD.enum textToScheduledEventStatus
+
+instance ToJSON ScheduledEventStatus where
+  toJSON = String . scheduledEventStatusToText
+
+instance FromJSON ScheduledEventStatus where
+  parseJSON = withText "String" $ \s ->
+    case textToScheduledEventStatus s of
+      Nothing     -> fail $ T.unpack $ "unexpected status: " <> s
+      Just status -> pure status
```

Comment:
```suggestion
  parseJSON = withText "String" $ \s -> onNothing (textToScheduledEventStatus s) $
    fail $ T.unpack $ "unexpected status: " <> s
```
(I didn't check how lambdas and `$` associate, so the above might need some parentheses before it compiles.)